Move to rust native errors for rust apis in CircuitData

mtreinish opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit migrates the CircuitData struct to have it's own new error enum CircuitDataError. This error type encapsulates all the possible error conditions returned from CircuitData methods and is used directly in rust. This has two goals, first it enables us to return real error messages to the C API, and secondly it decouples are rust code from Python as part of #14240. To enable a build without pyo3 we need to stop using PyErr in a pure rust context. The idea being that pymethods will be conditionally compiled and not built, but the other methods that aren't exposed to python will need to be returning rust error types. One thing to note is there is at least one method still using a PyResult, in those cases the methods will be conditionally compiled. Similarly there is an ErrorFromPython variant in the new CircuitDataError, this will also be conditionally compiled out, but is necessary because it's used in places where we are calling python explicitly and need to handle the error return from that.

The goal of the thiserror enum per module pattern is not to be a final arrangement, but instead as an incremental step to move us towards thinking of the error as part of the function's interface in rust. Right now most of the rust functions are treating errors like we do in Python where it can raise "an error" at any point. We should be moving towards a more defined model where the error type is explicitly part of the interface and has defined semantics that are understood by the caller. While we're transitioning our internal interfaces this enum model is meant to map all the errors we have internally into something explicitly defined as a baseline. This minimizes the amount of work needed to use the errors, but inventories all the error cases we have now. In the future we should move from this to explicit error types per function.

### Details and comments

Fixes #15027

Closes #15177
Closes #15062
Closes #15063

mtreinish committed:
Move to rust native errors for rust apis in CircuitData

This commit migrates the CircuitData struct to have it's own new error
enum CircuitDataError. This error type encapsulates all the possible
error conditions returned from CircuitData methods and is used directly
in rust. This has two goals, first it enables us to return real error
messages to the C API, and secondly it decouples are rust code from
Python as part of #14240. To enable a build without pyo3 we need to
stop using PyErr in a pure rust context. The idea being that pymethods
will be conditionally compiled and not built, but the other methods that
aren't exposed to python will need to be returning rust error types. One
thing to note is there is at least one method still using a PyResult,
in those cases the methods will be conditionally compiled. Similarly
there is an ErrorFromPython variant in the new CircuitDataError, this
will also be conditionally compiled out, but is necessary because it's
used in places where we are calling python explicitly and need to handle
the error return from that.

Fixes #15027

jakelishman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -46,9 +46,103 @@ use pyo3::{PyTraverseError, PyVisit, import_exception, intern};
+    #[error("invalid type for global phase")]
+    InvalidGlobalPhaseType,
+    #[error("{0}")]
+    ObjectRegistryError(ObjectRegistryError),
```
On Ray's PRs doing similar stuff to this, my instinct had been against these giant enums, since you're not really _supposed_ to handle it from Rust.

That said, something I hadn't previously considered properly was that the deferred conversion to a further type could be a goal in itself, and that if we've got tonnes of variants, it's probably less boilerplate than making new types for each of them.  This giant enum still doesn't _feel_ good to me, but I feel like I can get behind it as an intermediate step.

I guess the awkwardness is that the giant enum is so freeform is _feels_ `anyhow`-like in the "dump in any string" form so wants to be that just from Rust, but once we throw Python into the mix, we also want to have a way of specifying which exception type it should become, without having additional conditional code everywhere.

_Maybe_ an alternative would be us having
```
enum ErrorType {
  Value,
  Type,
  // ...
}
struct QiskitError {
  err: anyhow::Error,
  kind: ErrorType,
}
```
Just thinking it might cut down on the boilerplate, while still letting us defer the creation of the actual Python exception type?

(I'd have `QiskitErr::value`, `QiskitErr:type`, `...` constructors, rather than actually having to build the `ErrorType` thing)

mtreinish replied:
So I thought about this before settling on this structure and I liked this as a middle ground between having distinct types for every different method or different classes of errors and something like `anyhow`. One of the advantages here is that it reduces boiler plate on the conversion as you mentioned, and at the end of the day we're either returning this as a `PyErr`, converting it to something on the C side (error messages and return codes, etc), or handling it internally. I hear the argument against a big enum, but I actually like it as it encapsulates an enumeration of all the possible ways a struct/module can error in a single place. I think as we're moving to rust native errors we should be intentional on where we are returning errors and for what reasons. As a whole I think being more explicit about error conditions will make us more disciplined in thinking about it as part of the interface. 

I want to avoid `anyhow` except in maybe the C API layer because I thought that was way to free and similar to the python model of anything can error. I don't like your proposed `QiskitError` because it honestly feels a bit too close to what Python wants.  This model let us be explicit but also reduced the boiler plate when using the errors. 

That being said there is definitely an argument for finer or coarse grained error types though. I was scoping this per module was kind of going of how `std::io::ErrrorKind` and `std::io::Error` is used for `std::io` but using `thiserror` instead of doing it manually with a struct and enum field.

jakelishman replied:
> encapsulates an enumeration of all the possible ways a struct/module can error in a single place

This is actually why I'm against it - Rust-native error handling isn't done per module, it's done per function, so throwing everything into a huge enum largely removes the possibility of per-function handling, but the giant enum is still more discrete (implying boilerplate to me) than just `anyhow`. I don't really see this enum as anything other than freeform but with extra steps. If I wanted to throw a new error, my process would still be to just add stuff, but now I'd need to do it in more than one place.

That said, I'd rather see the PR merged in this current form than not merged at all.

mtreinish replied:
Per our offline discussion I think I made the intent a bit more clear in the PR summary and added docstrings to the error enums in: https://github.com/Qiskit/qiskit/pull/15279/commits/5bf6dab1288bcc55d05d7cfafc5cba3fd966513c with `[non_exhaustive]`

jakelishman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -2446,35 +2595,46 @@ impl CircuitData {
             };
-            Param::from_expr(new_expr, coerce)
+            // Param::from_expr() only errors in the python path when calling Python
+            Param::from_expr(new_expr, coerce).map_err(CircuitDataError::ErrorFromPython)
```
Feels like the method itself might want changing to reflect this?

mtreinish replied:
I probably should have been more precise, it's when the expression is a complex or integer value. Both of those aren't valid types on a `Param` without python (we just convert them to `Py<PyAny>` and store that. I will push up a PR adding int because that really needs to be in the data model, but we'll still have it for complex. In the case where it's a float value, an unbound expression, or an int or complex that is representable as a float the method is infallible.

jakelishman reviewed (commented):
Thanks for all the busywork on this.  I know we've already got quite far, I just wanted to see if maybe there's a less-boilerplate way to expand this in the future?

mtreinish committed:
Mark error enums as non-exhaustive and document their purpose

This commit adds the non_exhaustive attribute to all the error enums.
While we're using them there is the likelihood we'll add on to them with
new errors as methods are added. This also documents the enums are an
incremental step towards each function owning its error type as part of
the interface. These enums are just the first step towards that by
making it all explicitly defined. Eventually we'll want to split these
off into specific types that are specified per function/method instead
of being bundled together in a single enum.

debasmita2102 commented on crates/synthesis/src/multi_controlled/mcx.rs:
```diff
@@ -182,22 +167,25 @@ impl CircuitDataForSynthesis for CircuitData {
         if other.num_qubits() > self.num_qubits() {
-            return Err(QiskitError::new_err(
-                "Cannot compose a larger circuit onto a smaller circuit.",
-            ));
+            panic!("Cannot compose a larger circuit onto a smaller circuit.");
         }
```
I am wondering, as we migrate functions from `PyResult` to `CircuitDataError`, how do we determine which error conditions should map to new `CircuitDataError` variants versus those that should panic due to being true invariant violations? More specifically, were error cases like "Cannot compose a larger circuit onto a smaller circuit" intentionally left unmapped during this refactoring, or should they have been added to the `CircuitDataError` enum as part of completing the migration scope for this particular function? This would help me understand whether the `panic!` is a deliberate design choice or part of the incremental work still to be completed.

mtreinish replied:
In this case and most of this file it's a true invariant. In this case the compose() function is internal only, and it can't ever be called in this way. If it does trigger the panic the synthesis code is fundamentally broken. Arguably I should have just removed this check since it shouldn't be something we worry about inside multicontrolled x gate synthesis, but I didn't want to remove it to minimize the diff. We'll have to save that for a refactor/cleanup of the synthesis code in particular. A lot of this module in particular was returning `PyErr` unnecessarily for code that is fundamentally infallible and as I had to adapt for error returns I tried to clean some of it up.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into rust-error-circuit_data

mtreinish committed:
Merge remote-tracking branch 'origin/main' into rust-error-circuit_data

jakelishman reviewed (commented):
I've made mtreinish/qiskit-core#32, which is effectively my PR review of this, and fixes the merge conflict.  I'm happy to approve the HEAD of that PR.

jakelishman committed:
PR review of Qiskit/qiskit#15279 (#32)

* Add `thiserror` conversion traits

This uses the `thiserror::Error` derive-macro attributes
`#[error(transparent)]` and `#[from]` to remove boiler plate when
nesting other errors inside the `CircuitDataError` enum.  In particular,
adding a `#[from] PyErr` variant adds the relevant `From<PyErr>` trait
to `CircuitDataError` and makes the `?` operator work with chains of
Python calls again.

* Replace internal error with panic

An internal-logic error within Qiskit is a panic condition; if we don't
even know how we can get into that state, how can a user recover from
it?
